### ISSUE TITLE
material modifications cannot handle different values for a key within a given block

### ISSUE THREAD

**Issue by keckler** [terrapower/armi#513]
As described in the [docs on blueprints](https://terrapower.github.io/armi/user/inputs/blueprints.html?highlight=material%20modifications), the ability to specify material properties in a blueprint is typically through the use of material modifications, which are applied on the block level during assembly construction. 

Essentially what is done is that every component in a given block of the assembly tries to use all of the `matMods` keys, and the `matMods` that apply to the material class in each particular component end up being used while those that are irrelevant to the material class are ignored by the component. So if you had a configuration that looked like this (yes, I know it is weird to have Lithium as a coolant, just doing it to demonstrate):
```
blocks:
    fuel: &block_fuel
        fuel: &component_fuel_fuel
            shape: Circle
            material: UZr
            Tinput: 25.0
            Thot: 600.0
            id: 0.0
            mult: 169.0
            od: 0.86
        coolant: &component_fuel_coolant
            shape: DerivedShape
            material: Lithium
            Tinput: 450.0
            Thot: 450.0

assemblies:
    heights: &standard_heights [25.0]
    axial mesh points: &standard_axial_mesh_points [ 1]
    igniter fuel:
        specifier: IC
        blocks: &igniter_fuel_blocks [ *block_fuel]
        height: *standard_heights
        axial mesh points: *standard_axial_mesh_points
        material modifications:
            U235_wt_frac: [ 0.11]
            ZR_wt_frac: [0.06]
            LI_wt_frac: [0.5]
        xs types: [A]
```
then the UZr component would pick up the `U235_wt_frac` and `ZR_wt_frac` material modifications, while the Lithium coolant component would pick up the `LI_wt_frac` modification.

If you instead replaced the coolant component's material with UZr, then both components in the block would pick up the `U235_wt_frac` and `ZR_wt_frac` material modifications and neither would get the `LI_wt_frac`. And if the assembly were composed of multiple stacked blocks, one could specify the material modifications as a list to apply different values to the different blocks as in `ZR_wt_frac: [0.5, 0.6, 0.7]`.

However, right now there is no ability to specify unique enrichments for different materials that are _in the same block_. As an example, consider the following scenario where one has constructed a block with two different fuel components:
```
blocks:
    fuel: &block_fuel
        fuel1: &component_fuel_fuel1
            shape: Circle
            material: UZr
            Tinput: 25.0
            Thot: 600.0
            id: 0.0
            mult: 100
            od: 0.86
        fuel2: &component_fuel_fuel2
            shape: Circle
            material: UZr
            Tinput: 25.0
            Thot: 600.0
            id: 1.0
            mult: 69
            od: 1.5
        coolant: &component_fuel_coolant
            shape: DerivedShape
            material: Sodium
            Tinput: 450.0
            Thot: 450.0

assemblies:
    heights: &standard_heights [25.0]
    axial mesh points: &standard_axial_mesh_points [ 1]
    igniter fuel:
        specifier: IC
        blocks: &igniter_fuel_blocks [ *block_fuel]
        height: *standard_heights
        axial mesh points: *standard_axial_mesh_points
        material modifications:
            U235_wt_frac: [ 0.11]
            ZR_wt_frac: [0.06]
        xs types: [A]
```
In that case, both `fuel1` and `fuel2` get the same enrichments, though they are different designs and could very likely have different enrichment levels in real life. I do not believe there is a method of explicitly assigning each of these two components to have unique material modifications.

This is a barrier to using ARMI for many reactor designs. Not only does this situation occur in certain fast test reactors, but especially in LWRs or other thermal-spectrum, pin-based reactor designs, which often have fuel pins or burnable absorbers of varying enrichments at a given axial level in a single assembly.

I do not know exactly how this should be fixed, and it seems a little bit difficult with the existing architecture. Any thoughts on how to approach this would be welcome, and I will try to think about some viable solutions.

**Comment by keckler** [comment]
I am working on a fix for this.

**Comment by keckler** [comment]
Fixed with the closure of #517
